src/Components/Popover/Popover.tsx

```diff
@@ -0,0 +1,476 @@
+      return position
+    }
+
+    // Otherwise choose the next available position in from the acceptable list
+    return acceptablePopoverPositions[0] || null
+  }
+
+  private isDialogFlush = (position: PopoverPosition): PopoverFlush => {
+    const {triggerRect} = this.state
+
+    if (!triggerRect || !this.dialogRef.current) {
+      return {
+        first: false,
+        last: false,
+      }
+    }
+
+    const dialogRect = this.dialogRef.current.getBoundingClientRect()
+    // When the trigger is in a corner of the screen,
+    // determine whether to offest it
+    let first = false
+    let last = false
+
+    switch (position) {
+      case PopoverPosition.Above:
+      case PopoverPosition.Below:
+        // When the dialog is above or below the trigger
+        // First: left edge
+        // Last: right edge
+        const overflowX = dialogRect.width - triggerRect.width
+        first = triggerRect.left < overflowX / 2
+        last =
+          window.innerWidth - triggerRect.left - triggerRect.width <
+          overflowX / 2
+        break
+      case PopoverPosition.ToTheLeft:
+      case PopoverPosition.ToTheRight:
+        // When the dialog is left or right of the trigger
+        // First: top edge
+        // Last: bottom edge
+        const overflowY = dialogRect.height - triggerRect.height
+        first = triggerRect.top < overflowY / 2
+        last =
+          window.innerHeight - triggerRect.top - triggerRect.height <
+          overflowY / 2
+        break
+      default:
+        break
+    }
+
+    return {
+      first,
+      last,
+    }
+  }
+
+  private handleTriggerClick = (): void => {
+    const {showEvent} = this.props
+
+    if (showEvent === PopoverInteraction.Click) {
+      this.handleShowDialog()
+    }
+  }
+
+  private handleTriggerMouseOver = (): void => {
+    const {showEvent} = this.props
+
+    if (showEvent === PopoverInteraction.Hover) {
+      this.handleShowDialog()
+    }
+  }
+
+  private handleTriggerMouseOut = (): void => {
+    const {hideEvent} = this.props
+
+    if (hideEvent === PopoverInteraction.Hover) {
+      this.handleHideDialog()
+    }
+  }
+
+  private handleClickOutside = (): void => {
+    const {hideEvent} = this.props
+
+    if (hideEvent === PopoverInteraction.Click) {
+      this.handleHideDialog()
+    }
+  }
+
+  private handleUpdateTriggerRect = (): void => {
+    if (!this.triggerRef.current) {
+      return
+    }
+
+    this.setState({
+      triggerRect: this.triggerRef.current.getBoundingClientRect(),
+    })
+  }
+
+  private handleShowDialog = (): void => {
+    this.handleUpdateTriggerRect()
+
+    window.addEventListener('resize', this.handleUpdateTriggerRect)
+    window.addEventListener('scroll', this.handleUpdateTriggerRect)
+
+    this.setState({expanded: true})
+  }
+
+  private handleHideDialog = (): void => {
+    this.setState({expanded: false})
+
+    window.removeEventListener('resize', this.handleUpdateTriggerRect)
```

Comment:
Will this be fired when the component unmounts? It looks like it relies on a `mouseOut` event firing, but I'm not sure that will happen. Could be leaking memory.

Reply:
Good catch! I’ll add that in as a precaution